msg.c

```diff
@@ -170,11 +174,26 @@ int connect_monitor(char *devname)
 
 	addr.sun_family = PF_LOCAL;
 	strcpy(addr.sun_path, path);
-	if (connect(sfd, (struct sockaddr*)&addr, sizeof(addr)) < 0) {
-		close(sfd);
-		return -1;
-	}
 
+	/* In foreground mode, when mdadm is trying to connect to control
+	 * socket it is possible that the mdmon has not created it yet.
+	 * Give some time to mdmon to create socket.
+	 */
+	while (1) {
+		rv = connect(sfd, (struct sockaddr*)&addr, sizeof(addr));
+		if (rv < 0 && retry_count >= 10) {
+			pr_err("Failed to connect to control socket. (%s!!)\n",
+					strerror(errno));
+			close(sfd);
+			return -1;
+		} else if (rv < 0) {
+			retry_count++;
+			nanosleep(&delay, &delay);
```

Comment:
we have `sleep_for()` you can use.

Reply:
I wasn't sure about the 3rd parameter value, that's why I went with using nanosleep. You have used 3rd parameter (wake_after_interrupt) as false in this case. I believe that is the case in the scenario.

Reply:
Hmm, I looked into code and this parameter is set to true everywhere, also setting it to true will produce same result as as change you proposed (and tested). Please set it to true then, not false.